### ISSUE TITLE
Changing appearance of individual items

### ISSUE THREAD

**Issue by alongenosar** [danielsaidi/Sheeeeeeeeet#28]
How can i  change the appearance of individual items?
I tried this with no success
```
let a = ActionSheetItem(title: "foo", value: "1", image: nil )
a.appearance.textColor = .brown
```

**Comment by danielsaidi** [comment]
Hi!

I noticed that the action sheet always override the individual appearances of all its items every time it prepares to be presented.

I am working on a fix, which involves making the action sheet item `appearance` property optional. If you set it to a custom value, it will then not be overwritten by the action sheet appearance.

I'll let you know when it's done.

**Comment by danielsaidi** [comment]
Hi again @alongenosar and sorry for the slow reply. I have been very busy with work and life.

I started looking at this issue more than a week ago, but quickly ran into problems, since my initial idea about how item appearances work did not correlate with what I finally put into Sheeeeeeeeet. 

Let me try to explain.

An action sheet copies the global action sheet appearance when it is created. This means that any changes you've made to the global appearance will apply to all action sheets as they are created. After that, an action sheet's appearance is isolated from the global appearance. This means that you can then change the global appearance  without affecting the action sheet and vice versa.

However, the same does not apply for action sheet item appearances. Action sheet items will every now and then be asked to apply the appearance of their action sheet, e.g. every time the action sheet is refreshed. Every item type handles this a little differently, e.g. by copying different parts of the action sheet item appearance.

This means that the appearance property will be overwritten every now and then, which means that setting the property will have no effect.

I tried to make this possible, but ran into a clusterfudge of problems. It just doesn't work out. I think the way forward is to remove the `appearance` property to not give the impression that this property can be affected, and instead highlight the `applyAppearance` function instead, since this is where any customizations should be made.

So, to solve your acute problem, I think you subclass the item you want to use as template and override the `applyAppeance` function. Make sure to call `super`, then perform any changes you need.

Let me know if this works out for you.







All action sheet items in an action sheet are told to apply the action sheet's appearance multiple times, each time the sheet refreshes itself. The action sheet, on the other hand, copies the standard appearance o

**Comment by alongenosar** [comment]
@danielsaidi Hi. Thanks for taking the time to respond.  I eventually wrote my own simple action sheet from scratch that fits the project requirements. (see attached screenshot)
[https://ibb.co/kBS7uH](url)
Cheers
Alon

**Comment by danielsaidi** [comment]
Hi again,

Thanks for getting back with your own progress!

Ok, I'll reopen this just to make sure that I remember to hide the item appearance property or remove it altogether.

All the best!

**Comment by alongenosar** [comment]
I closed it by mistake. Sorry...

**Comment by danielsaidi** [comment]
Haha ok. Btw, the link you entered earlier is dead. I'm curious to see how your own sheet ended up.

**Comment by alongenosar** [comment]
Well Its a very simple action sheet. Attaching a screenshot and some code



  let facebook = ActionSheetButtonItem(title: "Facebook", icon:  imageLiteral(resourceName: "facebook_icon"), action: { print(“got action” })
  facebook.appearence.backgroundColor = UIColor.from(number: 0x4469b0ff)
  facebook.appearence.titleColor = UIColor.white
  facebook.appearence.imageEdgeInsets = UIEdgeInsetsMake(20, 0, 20, 0)
  facebook.appearence.height = 60
   
  let actionSheet = ActionSheet(withItems:items, callback:{ (item) in
           
  })
 
  actionSheet.present()

**Comment by danielsaidi** [comment]
I will have a look at this again and see if I can adjust the appearance handling for items. The code you just posted should be possible.

**Comment by digabriel** [comment]
+1

**Comment by alexandre-g** [comment]
Hi, as I understand `applyAppearance` workaround will not allow custom sized items?

**Comment by alexandre-g** [comment]
Hi Daniel, thanks for the quick reply.
Just to have it clear - the custom item would also be restricted to being the same height as other cells, right?

**Comment by danielsaidi** [comment]
The cell height will be correct for individual items, but the header view may be incorrectly positioned as the total height is incorrectly calculated.

**Comment by danielsaidi** [comment]
Hi again :) I just realized that the custom item adjusts its appearance copy’s height in apply(appearance: ...), which means that it will behave correctly. You could subclass any item and do the same and things will work.

The api should be redesigned, though :)

**Comment by alexandre-g** [comment]
Hey will give it a try, thanks. Btw, just pulled the latest version via Carthage and it's not building, giving
`No such module Quick`

**Comment by danielsaidi** [comment]
Hi!

I'll probably add a custom, optional height property to all items, so that you don't have to override the apply function. But please can try it out and see if it works for now :)

The build error is strange, since Travis can build it! Are you using Xcode 10 and Swift 4.2? Can you see if `Cartfile.private` is in the checkout? 

I'll give it a try in an hour, will try to fix it asap if I can reproduce the error.

**Comment by danielsaidi** [comment]
I also get errors when trying to install with Carthage. I'll look into it.

**Comment by alexandre-g** [comment]
Working thanks! That's how I fixed it for me, just wasn't sure that didn't break anything else

**Comment by danielsaidi** [comment]
Hi again @alongenosar @digabriel and @alexandre-g !

I think I have found a way to apply custom appearances for individual items now.

Basically, I added an optional `customAppearance` property to `ActionSheetItem`. It is`nil` by default. As `applyAppearance` is called on the item, it sets the `appearance` property to `customAppearance` if any. If not, it uses the provided appearance parameter, just like before.

The result of the following code in the example app's `HeaderActionSheet` class:

```
private extension HeaderActionSheet {
    
    static func items(for options: [FoodOption]) -> [ActionSheetItem] {
        var items = options.map { $0.item() }
        
        items[0].customAppearance = ActionSheetItemAppearance(copy: ActionSheetAppearance.standard.item)
        items[0].customAppearance?.textColor = .red
        items[0].customAppearance?.separatorInsets.left = 100
        
        items.insert(titleItem(title: standardTitle), at: 0)
        items.append(cancelButton)
        return items
    }
}
```

becomes:

![simulator screen shot - iphone xr - 2018-10-12 at 09 08 17](https://user-images.githubusercontent.com/429927/46853586-dabce580-cdfe-11e8-89c2-36e30d0ca5c4.png)

This will be available in `0.11.0`, so please try it out and let me know if it works as you'd like it to.

I still need to make copying another appearance instance easier. If I would have used `items[0].customAppearance = ActionSheetAppearance.standard.item`, I would still have affected the entire sheet.
